### PR TITLE
[MM-44667] Set InsecureSkipVerify for S3 when EnableInsecureOutgoingConnections is set

### PR DESCRIPTION
**Pull request by jakegut**
#### Summary

This PR will allow the S3 Client to connect to HTTPS servers with self-signed certificates when the `EnableInsecureOutgoingConnections` is set to true.

A tricky part of this PR was passing the `EnableInsecureOutgoingConnections` setting to the right place; open to suggestions if there's a better way to do it.

While I've added a test which checks that requests that are insecure are successful with self-signed HTTPS servers, I've to add a test that checks if the `EnableInsecureOutgoingConnections` setting enables insecure requests. Open to suggestions if I need to add that test.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/20356

#### Release Note

```release-note
Added insecure transport to S3 client when EnableInsecureOutgoingConnections is set to true.
```

### CODE REVIEW

File: shared/filestore/s3store.go [hunk 1]
```diff
@@ -115,10 +119,15 @@ func (b *S3FileBackend) s3New() (*s3.Client, error) {
 		creds = credentials.NewStatic(b.accessKey, b.secretKey, "", credentials.SignatureV4)
```
Comment (agnivade) [inline shared/filestore/s3store.go:119]: I was envisioning something like this. We just need to modify one field. `s3.DefaultTransport` already has some customizations:

```diff
 type S3FileBackendAuthError struct {
@@ -121,18 +123,22 @@ func (b *S3FileBackend) s3New() (*s3.Client, error) {
                Region: b.region,
        }

+       tr, err := s3.DefaultTransport(b.secure)
+       if err != nil {
+               return nil, err
+       }
+       if b.skipVerify {
+               tr.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+       }
+       opts.Transport = tr
+
        // If this is a cloud installation, we override the default transport.
        if isCloud {
-               tr, err := s3.DefaultTransport(b.secure)
-               if err != nil {
-                       return nil, err
-               }
                scheme := "http"
                if b.secure {
                        scheme = "https"
                }
                opts.Transport = &customTransport{
-                       base:   tr,
                        host:   b.endpoint,
                        scheme: scheme,
                }
```

Also, in `customTransport` we don't use the base at all. We can remove the field.

**Review by agnivade — COMMENTED** [review]
Great work on this! Just some minor questions on the tests.

File: shared/filestore/s3store_test.go [hunk 2]
```diff
@@ -77,3 +82,97 @@ func TestMakeBucket(t *testing.T) {
...
+				require.Error(t, err)
+			}
+		})
+	}
+}
+
+// adapted from https://gist.github.com/tjamet/8a3d9e3019045855137deadcf3318245
+func NewTLSProxyServer(client *http.Client, backend *url.URL) *httptest.Server {
+	return httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		req, err := http.NewRequest(r.Method, fmt.Sprintf("%s://%s", backend.Scheme, backend.Host), r.Body)
```
[the remark this answers is not in the document]
Comment (jakegut) [reply shared/filestore/s3store_test.go]: No reason. I found the `ReverseProxy` type, but I didn't see the `NewSingleHostReverseProxy` function until now.

**Review by agnivade — APPROVED** [review]
Fantastic work on this @jakegut !

**Review by streamer45 — APPROVED** [review]
Thank you @jakegut , looks great :+1:

**Comment by streamer45** [comment]
@agnivade Now that I think about it, there were breaking changes to `ToFileBackendSettings` so we'll have to update enterprise code as well.

**Comment by agnivade** [comment]
Yep, indeed. I'll close this and send another PR.

**Comment by agnivade** [comment]
EE PR here: https://github.com/mattermost/enterprise/pull/1240

**Comment by agnivade** [comment]
Strange, doesn't look like it worked.

**Comment by jakegut** [comment]
Closed because of https://github.com/mattermost/mattermost-server/commit/de50943d61adec1aa65ff30598c6e0893fa61bf9.

**Pull request closed without merging**